docs/api_specs/includes/watcher_api_specs.md

```diff
@@ -312,3 +312,110 @@ A block containing an invalid tx has been processed [are there other reasons a b
 
 The ChildChain is withholding a block whose hash has been published on the root chain. User should exit.
 
+#### `noncanonical_ife`
+> A noncanonical_ife event
+
+```json
+{
+    "event": "noncanonical_ife",
+    "details": {
+        "inflight_blocknum"  : 10000,
+        "inflight_tx_index"  : 12,
+        "inflight_input_index"  : 1,
+        "competing_blocknum"  : 10000,
+        "competing_tx_index"  : 10,
+        "competing_input_index"  : 0,
+    }
+}
+```
+
+An non-canonical in-flight exit has been started. It should be challenged. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Key | Type | Description
+--------- | ------- | -----------
+inflight_blocknum | integer | Block number of the in-flight transaction
+inflight_tx_index | integer | Index of the in-flight transaction
+inflight_input_index | integer | Index of the double-spent input in the in-flight transaction
+competing_blocknum | integer | Block number of the spending transaction
+competing_tx_index | integer | Index of the spending transaction
+competing_input_index | integer | Index of the double-spent input in the competing transaction
+
+#### `invalid_ife_challenge`
+> A invalid_ife_challenge event
+
+```json
+{
+    "event": "invalid_ife_challenge",
+    "details": {
+        "inflight_blocknum"  : 10000,
+        "inflight_tx_index"  : 12,
+    }
+}
+```
+
+A canonical in-flight exit has been challenged. The challenge should be responded to. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Key | Type | Description
+--------- | ------- | -----------
+inflight_blocknum | integer | Block number of the in-flight transaction
+inflight_tx_index | integer | Index of the in-flight transaction
+
+#### `piggyback_available`
+> A piggyback_available event
```

Comment:
1. Maybe `available_piggyback` to be consistent with `invalid_piggyback`.
2. We must not forget to specify endpoint which will be providing necessary information to make piggyback.